### PR TITLE
tests: unit tests for utils/scaffold-stark/contract.ts #373

### LINKED ISSUE #373: [TEST] Create unit tests for `utils/scaffold-stark/contract.ts` file
## Issue Overview

Unit tests required for the indicated file to ease future refactoring and feature validations.

## Proposed Solutions or Ideas (more like tips and reminders)

- We use vitest to run the unit tests, can refer to their documentation for the syntax: https://vitest.dev/, Jest developers shouldn't have too much of a hard time getting used to the syntax.
- Note that hooks testing tools are also now imported via `@testing-library/react`,
- Please refer to `README.md` on how to execute the tests
- Use `it.only()` to isolate tests in case you need to, but do not forget to remove it once you are submitting your PR.
- Mock when necessary. please make sure that the tests *NOT ONLY* works on your device
- If you're not sure of the features of the code unit to test, please refer to the [documentation](https://docs.scaffoldstark.com/hooks/), or free to ask on this issue page or on the Telegram Group

## Code Coverage
*IMPORTANT* - When submitting a PR, please provide a screenshot of the coverage for the specific units you are working on. We would like to achieve >= 70 % code coverage. 

Happy Hacking!

Example: 
image

### PR DESCRIPTION
**Pull request by od-hunter**
# Task name here

Fixes #373

## Types of change

- [ ] Feature
- [x] Bug
- [x] Enhancement

### CODE REVIEW

**Review by Nadai2010 — CHANGES_REQUESTED** [review]
Hi, thank you for your work on this!

- Please ensure you run `yarn format` to align the code with the formatting rules.
- The test path should be corrected to `utils/scaffold-start/__test__/contract.test.ts`. Make sure to include the `__test__` folder in the path.
- I'm encountering some error warnings related to the ABI (see the attached image for reference):  
  ABI Error Warnings
- Additionally, the test coverage is not at 100% (see the attached coverage report):  
  Coverage Report

File: packages/nextjs/utils/tests/contract.test.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      expect(result).toEqual([]);
+  });
+});
+
+describe('Contract Declarations', () => {
+  it('should check if contract declarations are missing', () => {
+      // Assuming contractsData is defined somewhere in the context
+      expect(contracts).toBeDefined();
+      
+  });
```
Comment (metalboyrick) [inline packages/nextjs/utils/tests/contract.test.ts:80]: not sure what is this case for

**Review by metalboyrick — CHANGES_REQUESTED** [review]
please address @Nadai2010 comments as well, they are important. 

On coverages cc @Nadai2010, not reaching 100% is fine but should be at least at 80%, currently i see 60%-ish coverage only. kindly add more case or test more possible branching (`if` statements, `switch` statements, optional chaining, etc) to increase this number.

Thanks for the hard work!

**Comment by metalboyrick** [comment]
Please attach a screenshot of the coverages as well @od-hunter

**Comment by od-hunter** [comment]
Tests passing
Screenshot from 2024-12-16 12-03-19
, though error warnings proving otherwise.

**Comment by od-hunter** [comment]
> Tests passing Screenshot from 2024-12-16 12-03-19 , though error warnings proving otherwise, resolving the type mismatch.

Error warnings fixed.

**Comment by od-hunter** [comment]
Screenshot from 2024-12-17 10-24-02\
The logic not covered are either private or functions logic in private functions.
Screenshot from 2024-12-17 10-27-02

**Comment by od-hunter** [comment]
@metalboyrick

File: packages/nextjs/utils/__test__/contract.test.ts
Comment (metalboyrick) [inline packages/nextjs/utils/__test__/contract.test.ts:1]: you forgot to delete this haha

**Review by metalboyrick — CHANGES_REQUESTED** [review]
you got this man, all the best!

File: packages/nextjs/utils/__test__/contract.test.ts [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+vi.mock("../contract", () => {
+  const originalModule = vi.importActual("../scaffold-stark/contract");
+  return {
+    ...originalModule,
+    // ...vi.importActual('../scaffold-stark/contract'),
+    parseParamWithType: vi.fn(),
+    encodeParamsWithType: vi.fn(),
+    decodeParamsWithType: vi.fn(),
+  };
+});
```
Comment (metalboyrick) [inline packages/nextjs/utils/__test__/contract.test.ts:20]: I don't think you should mock the stuff here, if you need to mock the functions on a one off basis, you can use the spy functions of vitest (https://vitest.dev/guide/mocking#functions). 

You can either play with the input or use the spy on the above to produce different cases, if some functions are private, sometimes it implies that you need to treat the logic inside as if they are part of the function without mocking.

**Review by metalboyrick — CHANGES_REQUESTED** [review]

**Comment by od-hunter** [comment]
@metalboyrick please review

**Review by metalboyrick — APPROVED** [review]
looks good to me ! let's have @Nadai2010 review again to see if his issues are resolved, thanks for the hard work man!

**Review by Nadai2010 — CHANGES_REQUESTED** [review]
For me it's all good, we just need to move to the correct route that we discussed here. (2 point)

https://github.com/Scaffold-Stark/scaffold-stark-2/pull/391#pullrequestreview-2504244568

imagen

**Comment by od-hunter** [comment]
@Nadai2010 , please review

**Review by Nadai2010 — APPROVED** [review]
Thank you for your work

**Pull request merged**